mlflow/pyspark/ml/__init__.py

```diff
@@ -643,19 +834,73 @@ def fit_mlflow(original, self, *args, **kwargs):
             _logger.warning(_get_warning_msg_for_fit_call_with_a_list_of_params(self))
             return original(self, *args, **kwargs)
         else:
-            estimator = self.copy(params)
+            estimator = self.copy(params) if params is not None else self
             _log_pretraining_metadata(estimator, params)
             spark_model = original(self, *args, **kwargs)
             _log_posttraining_metadata(estimator, spark_model, params)
             return spark_model
 
     def patched_fit(original, self, *args, **kwargs):
+
+        status = _get_autologging_metrics_manager()
+        should_log_post_training_metrics = status.should_log_post_training_metrics()
         with _SparkTrainingSession(clazz=self.__class__, allow_children=False) as t:
             if t.should_log():
-                return fit_mlflow(original, self, *args, **kwargs)
+                with status.disable_log_post_training_metrics():
+                    return fit_mlflow(original, self, *args, **kwargs)
+                if should_log_post_training_metrics:
+                    status.register_model(self, mlflow.active_run().info.run_id)
             else:
                 return original(self, *args, **kwargs)
 
+    def patched_transform(original, self, *args, **kwargs):
+        status = _get_autologging_metrics_manager()
+        if status.should_log_post_training_metrics() and status.get_run_id_for_model(self):
+            predict_result = original(self, *args, **kwargs)
+            eval_dataset = get_instance_method_arg_value(0, "dataset", None, args, kwargs)
+            eval_dataset_name = status.register_prediction_input_dataset(self, eval_dataset)
+            status.register_prediction_result(
+                status.get_run_id_for_model(self), eval_dataset_name, predict_result
+            )
+            return predict_result
+        else:
+            return original(self, *args, **kwargs)
+
+    def patched_evaluate(original, self, *args, **kwargs):
+        status = _get_autologging_metrics_manager()
+        if status.should_log_post_training_metrics():
+            with status.disable_log_post_training_metrics():
+                metric = original(self, *args, **kwargs)
+
+            if status.is_metric_value_loggable(metric):
+
+                metric_name = self.getMetricName()
+
+                params = get_instance_method_arg_value(1, "params", None, args, kwargs)
+                evaluator = self.copy(params) if params is not None else self
```

Comment:
Can you remind me why we perform this copying after producing the eval metric? Perhaps an inline comment explaining why this is necessary would help.

Reply:
https://spark.apache.org/docs/latest/api/python/_modules/pyspark/ml/param.html#Params.copy

```python
    def copy(self, extra=None):
        """
        Creates a copy of this instance with the same uid and some
        extra params. The default implementation creates a
        shallow copy using :py:func:`copy.copy`, and then copies the
        embedded and extra parameters over and returns the copy.
        Subclasses should override this method if the default approach
        is not sufficient.

        Parameters
        ----------
        extra : dict, optional
            Extra parameters to copy to the new instance

        Returns
        -------
        :py:class:`Params`
            Copy of this instance
        """
        if extra is None:
            extra = dict()
        that = copy.copy(self)
        that._paramMap = {}
        that._defaultParamMap = {}
        return self._copyValues(that, extra)
```

Reply:
yes. because we need generate evalautor params here. after self.copy(params) we can get the correct param maps from `evalautor.params`